### ISSUE TITLE
[network] re-thinking network peer id story

### ISSUE THREAD

**Issue by phlip9** [libra/libra#3960]
### Current State

LibraNet was originally written for the validator network use-case, while other use-cases such as the validator-internal network and Public FullNode network were engineered in afterwards. These historical artefacts are apparent in how we identify peers in LibraNet networks.

All connections across all use-cases are authenticated (either server-only or mutually) and encrypted with NoiseIK. The NoiseIK handshake proves ownership of a network identity keypair (specifically, an `(x25519::PrivateKey, x25519::PublicKey)` keypair).

In the original validator network use-case, connections are identified and dedup'ed by each validator peers' on-chain `AccountAddress`. In a connection handshake, the validator peer's `x25519::PublicKey` is reverse-mapped to the validator's `AccountAddress` via the on-chain `ValidatorSet`, which for our purposes we can abstract as a `HashMap<AccountAddress, Vec<x25519::PublicKey>>`. This fact is important because it allows us to maintain _at most one_ connection per validator and simplify application protocols, since the mental model is just one validator, one account address, one peer id, one connection.

For validator-internal network and public fullnode network use-cases, `PeerId = AccountAddress = Hash(x25519::PublicKey)`, which is awkward as `AccountAddress` is meaningless in these cases. In an ideal world, `PeerId` would just be a `x25519::PublicKey`. The intention of this issue is to explore how we can remove this weird conversion or otherwise rethink network id's as whole in LibraNet.

### Options

1. Do nothing. `PeerId = AccountAddress` everywhere.

2. Make `PeerId` generic, e.g., `NetworkEvents<TPeerId, TMessage>`.

For validator networks, `PeerId = AccountAddress` while in others `PeerId = x25519::PublicKey`. Validator transports would probably have a `.map(|connection| { .. })` that maps negotiated `x25519::PublicKey` to `AccountAddress`. Unfortunately, adding a generic `PeerId` everywhere in `network/` would cause significant API churn and complicate application protocols like state sync and mempool, which need to hold multiple distinct network handles (which would now be different types).

3. `PeerId = x25519::PublicKey` everywhere.

In this world, LibraNet is effectively unaware of validators as a special case. The key part to explore here is how this changes existing validator network semantics. For example, without further changes, this would subtly break simultaneous dial tie breaking and so we could possibly hold multiple connections open to/from a single validator (at most one connection for each of their advertised pubkeys). Connectivity Manager could try to dedup validator connections, but this would be somewhat messy and racy (since other application protocols would get notified of the new connection simultaneously). This might also confuse some application protocols like consensus, which expect 1 validator = 1 peer id = 1 connection. For inbound messages, consensus would probably have to reverse map `x25519::PublicKey` -> `AccountAddress` like we do in network today. For outbound messages, consensus has to choose which connected pubkey to send the message to. Messages might be reordered if the outbound pubkey changes.

4. Suggestions?

**Comment by mimoo** [comment]
can a validator have multiple network public keys? I know this was considered earlier, but not anymore right? In this case I don't think option 2 really adds much as would have `PeerId = x25519::PublicKey` in the validator network too

**Comment by phlip9** [comment]
the node software doesn't support multiple network keypairs today, but there's nothing stopping us from advertising multiple pubkeys or listening with multiple pubkeys (we will already need to do this for rotations). however, at least for v1, we probably won't support automated rotations with multiple advertised pubkeys. instead, we'll just do one-step rotation on the same port b/c it's much simpler.

1. `advertised_addrs: [ "/../tcp/123/ln-noise-ik/<pk1>/.." ]`
2. (rotate to `<pk2>`)
3. `advertised_addrs: [ "/../tcp/123/ln-noise-ik/<pk2>/.." ]`

anyway, what i'm trying to say is that we shouldn't restrict a validator from advertising multiple pubkeys:

`advertised_addrs: [ "/../ln-noise-ik/<pk1>/..", "/../ln-noise-ik/<pk2>/..", .. ]`

**Comment by davidiw** [comment]
This is a hard problem, I suspect we shouldn't invest too much time in it prior to V1 and just accept what we have ... I lean toward #1 for now...

**Comment by mimoo** [comment]
I don't think it's a hard problem, but it's also not a priority yes. Let's keep discussing for now.

**Comment by davidiw** [comment]
At @mimoo, this is up to the noise key exchange to facilitate. I think we discussed this last week and came to this conclusion:

TrustedPeers can allow arbitrary mapping. From our perspective, this means that valdiators and fullnodes on chain should use the account of validator (owner) as the peer id for networking.

For "unauthenticated peers", we need a way to deterministically compute their peer id, so that they can confuse networking about other peers. For these peers, namely those not in TrustedPeers, we derive their account from their x25519 key.

This is good enough for V1, in the future, we could explore certification of ids or some globally trusted peer id pki, but sooo out of scope for V1.

**Comment by phlip9** [comment]
The awkward part is that in "client-only" auth networks, the dialing peer's derived `PeerId = x25519::PublicKey[16..32]` which introduces some subtle changes to the security model. Namely, the dialer is proving ownership of a 16-byte suffix of a public key (instead of the entire public key), since pubkeys with the same suffix but different prefix are considered to be the same peer.

Of course, this is still probably Good Enough™ for now, but it would be nice to eventually make the security model sensible again.

As far as timelines, I think we're all in agreement that this is post V1 work.

**Comment by mimoo** [comment]
This shouldn't matter, as you wouldn't be able to generate a public key that has 16 bytes that collide with another peer.

> his means that valdiators and fullnodes on chain should use the account of validator (owner) as the peer id for networking.

@davidiw that's what we do (at least in @phlip9 's monolitic PR)

**Comment by davidiw** [comment]
Awesome, I'm glad we're all in consensus. Though I do look forward to discussing our fullnode ecosystem post  V1.

**Comment by davidiw** [comment]
> This shouldn't matter, as you wouldn't be able to generate a public key that has 16 bytes that collide with another peer.
> 
> > his means that valdiators and fullnodes on chain should use the account of validator (owner) as the peer id for networking.
> 
> @davidiw that's what we do (at least in @phlip9 's monolitic PR)

I don't think we ask for the PeerId, we just derive it. I want us to explicitly ask for it to mitigate id confusion.  I know for a fact that in my tests, my nodes were not using the Id I intended but were still able to authenticate because the remote node had a different PeerId for them then they had locally. Oh the madness.

**Comment by davidiw** [comment]
Per our discussion, during genesis I do everything correctly and I initialize my validator to run as account {{alice}}, but in my node config I accidentally set my account to {{bob}}. Later on another validator operator who manages {{carol}} says that {{alice}} is flooding consensus with proposals. Everyone checks their configs to see who {{alice}} is but no one can find it.

Now in a validator network, this is a bit harder to happen because our deployment derives this and has been verified to be correct, but if folks don't use our management tool and do everything by hand there is nothing to ensure that this is correct. Hence we really want to make sure that {{alice}} and {{carol}} agree that they are talking to {{alice}} and {{carol}} and that there is no {{bob}}.

**Comment by mimoo** [comment]
> Everyone checks their configs to see who {{alice}} is but no one can find it.

How is this possible though? There's a validator key signing proposals and nobody knows who that is?

**Comment by davidiw** [comment]
hahaha, not a perfect example, I think the validator network will be harder to screw up... but as you can see in this commit: I fixed a case where I was using wrong IDs: https://github.com/libra/libra/pull/4014/commits/e7368033d89941b250c09314596a7dceb39a726b

the reality is if we allow folks to manually configure their Id, logs could be a mess... it is better to ensure we have a consistent view point of the world

**Comment by davidiw** [comment]
> > Everyone checks their configs to see who {{alice}} is but no one can find it.
> 
> How is this possible though? There's a validator key signing proposals and nobody knows who that is? That seems like a consensus issue not a network issue.

Consensus derives identity from network, which derives it locally from the config and externally from the trusted peers.

**Comment by mimoo** [comment]
interesting, I'll take a look at your example. I'm wondering if this can be fixed by not having your peerid in a config (only public keys)

**Comment by davidiw** [comment]
You cannot force derivation, that's the problem. A validator and vfn get their ids from the validator owner's account. The owner's account is initially mapped to the owner's public key -- until it is rotated. Chaos!
